add initial, basic grafana_annotation resource

mdb opened the pull request:
This seeks to address issue #11

Note that, because Terraform manages the CRUD lifecycle of resources declared in `*.tf` configuration, the `grafana_annotation` resource creates an annotation on initial `terraform apply`, then continuously updates that _same_ annotation (or deletes that same annotation if the `grafana_annotation` declaration is removed from `*.tf` configuration). This may be unexpected for those users who expect `grafana_annotation` to create fresh annotations with each `terraform apply` invocation and to leave in place those annotations created by preceding `terraform apply` invocations.

mdb committed:
add initial, basic grafana_annotation resource

This seeks to address issue #11

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
appease golangci-lint

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
convert set to string slice

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
error if API returns no annotations

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
correct annotation destroy check test logic

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
test grafana_annotation update scenario

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb commented on grafana/resource_annotation_test.go:
```diff
@@ -0,0 +1,109 @@
+				ResourceName:      "grafana_annotation.test",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
```
Are there additional test cases worth adding?

julienduchesne replied:
Should be enough

julienduchesne commented on grafana/resource_annotation.go:
```diff
@@ -0,0 +1,174 @@
+			"panel_id": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     false,
```
Should be required
```suggestion
				Required:    true,
```

mdb replied:
@julienduchesne As mentioned above, the [Grafana API docs](https://grafana.com/docs/grafana/latest/developers/http_api/annotations/#create-annotation) lead me to believe `panel_id` should be optional.

julienduchesne replied:
Indeed, you're right. Then the default value should just be empty, not `false`. A boolean doesn't make much sense in this context. Also, 
- it probably needs to cause a recreate to change the `panelId` or the `dashboardId`
- I think panelId is required when you set dashboardId and vice-versa. You can use `RequiredWith` for this
- It needs a test case, we need to test that having an annotation without panelId and setting one does recreate the resource correctly

mdb replied:
Thanks for the feedback. Stay tuned; I can cook all that up...

mdb replied:
> I think panelId is required when you set dashboardId and vice-versa.

...and I do think we need to verify this, as this has not been my understanding, historically.

mdb replied:
> it probably needs to cause a recreate to change the `panelId` or the `dashboardId`

@julienduchesne Is this to say you'd like to see a `ForceNew: true` associated with the `dashboard_id` and `panel_id` `*schema.Schema`? If so, I've added this.

> I think panelId is required when you set dashboardId and vice-versa.

@julienduchesne I don't believe this is true; I don't believe `panelId` and `dashboardId` are co-dependent, nor are either required. I believe I've proven this via an acceptance test configuration that uses `dashboard_id` without `panel_id`, and an acceptance test configuration that uses `panel_id` without a `dashboard_id`. Do you agree?

> It needs a test case, we need to test that having an annotation without panelId and setting one does recreate the resource correctly

@julienduchesne I believe I've added test cases covering these scenarios. Are there other test cases you'd like to see?

julienduchesne commented on grafana/resource_annotation.go:
```diff
@@ -0,0 +1,174 @@
+			"dashboard_id": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Default:     false,
```
Should be required
```suggestion
				Required:    true,
```

mdb replied:
@julienduchesne Are you certain `dashboard_id` should be required? According to the [Grafana API docs](https://grafana.com/docs/grafana/latest/developers/http_api/annotations/#create-annotation):

> The `dashboardId` and `panelId` fields are optional. If they are not specified then an organization annotation is created and can be queried in any dashboard that adds the Grafana annotations data source.

mdb replied:
I don't believe `dashboard_id` should be required.

julienduchesne commented on grafana/resource_annotation.go:
```diff
@@ -0,0 +1,174 @@
+			"text": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     false,
```
Should be required
```suggestion
				Required:    true,
```

julienduchesne commented on docs/resources/annotation.md:
```diff
@@ -0,0 +1,36 @@
+- `dashboard_id` (Number) The ID of the dashboard on which to create the annotation. Defaults to `false`.
+- `panel_id` (Number) The ID of the dashboard panel on which to create the annotation. Defaults to `false`.
+- `tags` (Set of String) The tags to associate with the annotation.
+- `text` (String) The text to associate with the annotation. Defaults to `false`.
```
I think having the time and timeEnd attributes is needed for a MVP. If we're managing annotations with Terraform, they're probably historical annotations, not ones that apply to right now

julienduchesne commented on grafana/resource_annotation_test.go:
```diff
@@ -0,0 +1,109 @@
+		}
+
+		for _, annotation := range annotations {
+			if annotation.Text == testAccAnnotationInitialText || annotation.Text == testAccAnnotationUpdatedText {
```
Since only this resource will ever create annotations and tests do not run in parallel, we could just fail whenever there's an annotation, no matter the text. It'd be future-test proof (no need to keep the test up-to-date)

julienduchesne reviewed (commented):
Great work. It's almost ready, just missing some attributes 🙂

mdb committed:
grafana annotation text is required

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
simplify annotation resource test logic

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
build trigger

mdb committed:
remove incompatible default setting

This seeks to fix...

```
* resource grafana_annotation: text: Default cannot be set with Required
```

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
fix docs drift

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
remove copy/pasta defaults

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
update annotation documentation

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
add time/time_end arguments to annotation

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
update annotation documentation w/ time/time_end

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
acceptance test annotation w/ dashboard_id

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
force new annotation w/ dashboard_id/panel_id change

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
uniquely/consistently name annotation test resources

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
acceptance test annotation w/ panel_id argument

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
use data source to retrieve dashboard JSON

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
attempt to correct JSON parsing error

This seeks to correct...

```
=== RUN   TestAccAnnotation_basic

    resource_annotation_test.go:24: Step 7/9 error: Error running apply: exit status 1

        Error: Unsupported attribute

          on terraform_plugin_test.tf line 19, in locals:

          19:   panel_id       = local.dashboard_json.dashboard.panels[0].id

            ├────────────────

            │ local.dashboard_json is object with 5 attributes

        This object does not have an attribute named "dashboard".
```

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
explicitly set panel ID in annotation test

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
remove use of data source in annotion test

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
Revert "remove use of data source in annotion test"

This reverts commit 969d36ab5ffd44734a08cc7349dcf66cbb37aad7.

mdb committed:
build trigger

mdb committed:
build trigger

julienduchesne reviewed (approved):
LGTM! Thanks!
